PCRelate improvements

jbloom22 opened the pull request:
Still in progress, picking up from #3193.

jbloom22 commented:
@konradjk I'm picking this up from this PR and closed the other one

konradjk commented on python/hail/methods/statgen.py:
```diff
@@ -1719,25 +1717,106 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+
+    col_keys_table = ds.col_keys()
+    col_keys_local = col_keys_table.collect()
+    col_keys_schema = col_keys_table.schema
```
This is the next line to fix, right? `.row` is certainly more correct, but not sure how to get `_convert_to_j` from there)

tpoterba replied:
if the scala method returns i,j ints then we can collect + broadcast and everything is peachy.

konradjk commented on python/hail/methods/statgen.py:
```diff
@@ -1468,16 +1467,16 @@ def pca(entry_expr, k=10, compute_loadings=False, as_array=False):
-    """Compute relatedness estimates between individuals using a variant
-    of the PC-Relate method.
+           statistics=enumeration('phi', 'phik2', 'phik2k0', 'all'))
+def pc_relate(mt,  maf, k=None, scores=None, block_size=None, min_kinship=-float("inf"), statistics="all"):
```
Extra space before `maf`

konradjk commented on python/hail/methods/statgen.py:
```diff
@@ -1690,54 +1678,92 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+    k : :obj:`int`, optional
         The number of principal components to use to distinguish ancestries.
+        Exactly one of `k` and `scores` must be specified.
+    scores : :class:``, optional
```
Missing `Table`

konradjk commented:
To be clear, I'm approving the Python code, and the fact that it runs to completion (will check for correctness shortly). Someone else might want to look over the Scala code changes)

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1688,56 +1677,105 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
-        :py:data:`.tfloat64`, `k1`: :py:data:`.tfloat64`, `k0`:
-        :py:data:`.tfloat64`. The table is keyed by `i` and `j`.
+    scores_local = scores_table.collect()
+    col_key_java = list(map(lambda x: ck_type._convert_to_j(x[ck_name]), scores_local))
```
We can't add more convert_to_j calls. We want to delete these methods, not proliferate them.

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1466,46 +1465,60 @@ def pca(entry_expr, k=10, compute_loadings=False, as_array=False):
 
     .. include:: ../_templates/req_biallelic.rst
 
+    .. include:: ../_templates/req_tstring.rst
```
this should not require strings.

danking replied:
I thought I re-wrote this to avoid strings, is there somewhere I forgot to do that?

jbloom22 replied:
The generalization was causing problems, so I pulled it back to get to something that worked. I'll check again if it's simple to re-generalize (I put FIXME), trying to keep this PR to a merge-able improvement that includes Konrad's use case.

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1466,46 +1465,60 @@ def pca(entry_expr, k=10, compute_loadings=False, as_array=False):
-    of the PC-Relate method.
+           statistics=enumeration('phi', 'phik2', 'phik2k0', 'all'),
+           block_size=nullable(int))
+def pc_relate(mt, maf, k=None, scores=None, min_kinship=-float("inf"), statistics="all", block_size=None):
```
if k and scores are mutually exclusive arguments, make them keyword only arguments:
```
def pc_relate(mt, maf, min_kinship=-float("inf"), statistics="all", block_size=None, *, k=None, scores=None): ```

danking replied:
@tpoterba what's the pythonista feeling on one argument that takes either an integer or a table? We'd need a new `either` type checker. I kind of like:

```
pc_relate(mt, mt.aaf, scores=5)
pc_realte(mt, mt.aaf, scores=mt.pc_score)
```

danking replied:
Erm I guess I don't want a table, I really want a table-like expression. Unrelated issue though.

tpoterba replied:
That seems OK too. Won't `oneof` work here?

konradjk replied:
Hrm, I'd suggest against. The user really should really be explicit about what they're putting in there.

konradjk replied:
Also `scores=5` doesn't really make sense. It really is `k`

danking replied:
langs really need to do a better job supporting mutually exclusive arguments, imo.

konradjk replied:
No argument (heh) there. But I don't think it's so bad if they're documented next to each other, and as mutually exclusive.

jbloom22 replied:
So I'll just make the `*, k=None, scores=None` change to force keywords?

jbloom22 replied:
Actually, I prefer this, keeping the more commonly used parameters toward the front:
`pc_relate(mt, maf, *, k=None, scores=None, min_kinship=-float("inf"), statistics="all", block_size=None):`

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1688,56 +1677,105 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
-                   intstatistics))
+        .apply(Env.hc()._jhc,
+               g._jbm,
+               jarray(Env.jvm().java.lang.Object, col_key_java),  # FIXME this should be avoidable
```
this is avoidable by using integer column indices and mapping back the true keys after.

jbloom22 commented:
I now pass the scores_table through as a Table rather than localizing and passing through colKeys, colKeyType, and scores annotations. The column key can now be any type. Both string and integer keys are tested from Python.

However, `requireUniqueSamples` still requires a single string ID (this was the remaining problem of going generic), so I've removed this check and would appreciate feedback on the best approach to checking uniqueness, preferably on the localized `keys` in PCRelate so as not to trigger additional actions.  I could use keyType.valuesSimilar to compare any two elements...it's a bit weird to have a tolerance on floats here.

As noted, I'm also a bit wary that I'm relying on `scores` from `pca` to be in the same order as the columns on the matrix table. This is currently true, but could change.

@danking I think the joins in `fuse` should also be zipPartitions, I've noted it in a FIXME.

tpoterba commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -29,34 +33,173 @@ object PCRelate {
   val defaultStatisticSubset: StatisticSubset = PhiK2K0K1
+  val defaultStorageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK
+
+  private[this] def apply(
```
there are 7 overloaded methods here. Can we reduce this to 2 (one for Scala and one for Python)?

tpoterba commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -29,34 +33,173 @@ object PCRelate {
+    val scoresLocal = scoresTable.collect()
+    assert(scoresLocal.length == blockedG.nCols)
+
+    val keyQueriers = scoresTable.key.map(scoresTable.signature.query(_))        
```
I think I proposed the wrong thing -- Instead, we can annotate the MT in Python and just pass in a field name for the PCs.

tpoterba commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -29,34 +33,173 @@ object PCRelate {
+    
+    apply(vds.hc,
+      blockedG,
+      vds.stringSampleIds.toArray[Annotation],
```
Can't we delete every signature that takes an array of annotations and just use integer indices?

tpoterba commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -29,34 +33,173 @@ object PCRelate {
+    minKinship: Double,
+    statistics: PCRelate.StatisticSubset): Table = {
+    
+    val scoresLocal = scoresTable.collect()
```
we definitely can't rely on these being the same keys and in the same order.

jbloom22 commented:
- Method now takes expressions for call and (optionally) scores.
- Block matrix and table of scores annotated and collected from source.cols() sent to Python, processed using int indices, column names restored on python side (thanks @tpoterba)
- Fixed bug that silently dropped `n_samples / block_size` proportion of pairs, Python test checks it
- Extended Python tests to compare k and scores paths, test counts, min_kinship, maf, block_size
- Tuned tolerances on comparison with R from Python
- Extended to general column key, removing unique key check, noted in docs
- MEMORY_AND_DISK caching as default (thanks @konradjk) on Scala side
- The diagonal fix meant phi is computed with parallelism up to the number of diagonal blocks, rathe than 1. But that's still likely a bottleneck as phi requires computing and point-wise dividing two big gram matrices. I now write phi to disk and read it back in, which squares the parallelism up to the number of blocks in phi. I think this should also improve the stability of the many downstream calculations derived from phi, esp. if pre-emptibles are used. No longer cacheing phi, but I left it on the other matrices.
- Noted in FIXME room for further improvement when fusing blocks: `replace join with zipPartitions, throw away lower triangular blocks sooner, avoid the nulls`
- Updated docs accordingly

jbloom22 commented:
And R packages fixed up on CI, thanks @danking !

jbloom22 commented:
Added check that all columns have score arrays (in Python), that all score arrays have same length, and that all score array values is non-missing (in Scala). Later, I think we ought to build conversion of a Table of Arrays to a MatrixTable of entries, and then we can convert the score table to a block matrix in Python.

tpoterba commented on python/hail/methods/misc.py:
```diff
@@ -155,6 +157,10 @@ def require_biallelic(dataset, method) -> MatrixTable:
     return dataset
 
+@typecheck(ds=MatrixTable, method_name=str)
+def require_unique_samples(ds, method_name):
```
delete

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1350,18 +1350,17 @@ def hwe_normalized_pca(dataset, k=10, compute_loadings=False, as_array=False) ->
-                 k,
-                 compute_loadings,
-                 as_array)
+    dataset = dataset.annotate_rows(mean_gt = dataset.AC / dataset.n_called)
```
prefix with __ ( and above)

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1467,46 +1466,57 @@ def pca(entry_expr, k=10, compute_loadings=False, as_array=False) -> Tuple[List[
-    of the PC-Relate method.
+           statistics=enumeration('kin', 'kink2', 'kink2k0', 'all'),
+           block_size=nullable(int))
+def pc_relate(call_expr, maf, *, k=None, scores_expr=None,
```
can we rename maf to min_maf

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1689,57 +1675,114 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
-        The number of principal components to use to distinguish ancestries.
+    call_expr : :class:`.CallExpression`
+        Entry-indexed call expression.
     maf : :obj:`float`
```
rename to min_individual_af

konradjk replied:
Yeah it's thrown me off a few times that this is not a cohort MAF, but individual. Would be good to make that distinction so people don't think they filtered their dataset more than they did

jbloom22 replied:
`min_individual_maf` since it's minor allele freq

tpoterba commented on python/hail/methods/statgen.py:
```diff
@@ -1689,57 +1675,114 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+    if n_missing > 0:
+        raise ValueError(f'Found {n_missing} columns with missing scores array.')
+
+    mt = mt.select_entries(gt=call_expr.n_alt_alleles())
```
__gt

danking commented on python/hail/methods/statgen.py:
```diff
@@ -1627,19 +1638,14 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
-    :func:`.pc_relate` differs from the reference
-    implementation in a couple key ways:
-
-     - the principal components analysis does not use an unrelated set of
```
as discussed offline I think this needs to be preserved

danking commented on python/hail/methods/statgen.py:
```diff
@@ -1689,57 +1676,119 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+     - `kin` (:py:data:`.tfloat64`) -- Kinship estimate, :math:`\\widehat{\\phi_{ij}}`.
+     - `k2` (:py:data:`.tfloat64`) -- IBD2 estimate, :math:`\\widehat{k^{(2)}_{ij}}`.
+     - `k0` (:py:data:`.tfloat64`) -- IBD1 estimate, :math:`\\widehat{k^{(0)}_{ij}}`.
+     - `k1` (:py:data:`.tfloat64`) -- IBD0 estimate, :math:`\\widehat{k^{(1)}_{ij}}`.
```
switch "IBD1 estimate" and "IBD0 estimate"

danking replied:
It's a bit weird that we now have three of these referring to the mathematical notation and one of them referring to the english language description, but I don't really care much.

jbloom22 replied:
updated

danking commented on python/hail/methods/statgen.py:
```diff
@@ -2144,10 +2193,9 @@ def genetic_relatedness_matrix(dataset) -> KinshipMatrix:
 
     return KinshipMatrix._from_block_matrix(tstr,
                                             grm,
-                                            [row.s for row in dataset.cols().select('s').collect()],
```
This should really be the whole col_key, right?

danking replied:
yes for this method, `genetic_relatedness_matrix`.

danking replied:
I think we all agree it should get fixed, but not in this PR.

jbloom22 replied:
Yes for GRM, but once this PR goes in, you can run GRM so long as the column key is a single string. RRM already has this change.

danking commented on src/main/scala/is/hail/methods/PCRelate.scala:
```diff
@@ -29,62 +30,124 @@ object PCRelate {
+        var jj = 0
         while (jj < lmPhi.cols) {
           var ii = 0
-          // assumes square blocks
```
this still assumes square blocks because of the calculation of `rowsAboveDiagonal`, right?
